리플렉션 실습 2단계 - 동적 빈 생성

kangminhyuk1111 opened the pull request:
1단계 PR 요청 피드백 감사합니다.

2단계 내용 PR 올렸습니다.

궁금한점
- 1번째 요구사항에 특정 애너테이션이 붙은 클래스만 가져오기 이 부분을 구현할 때, 어노테이션이 총 5개가 있는데 모든 어노테이션 생성해서 푸는게 맞을까요?

2번째 요구사항 - 조회한 클래스들을 통해 빈 생성에 대한 고민 및 해결 과정
1번째 요구사항에서 제시된 어노테이션들을 가져오는 것 까지 했으나 요구 사항 2의 문제를 해결할때 싱글톤 객체로 만들어야 해야하기 때문에 어떻게 해야 될지 방법을 찾아야 했습니다.
구글링 해서 찾아본 결과 싱글톤이란 간단하게 특정 클래스의 인스턴스는 하나만 존재하도록 하는 디자인 패턴이라고 이해했습니다.
고민한 해결 방법은 생성한 인스턴스가 존재하는지 검증을 하고 컬렉션에 추가한다면 항상 하나의 인스턴스만 존재할것이다 라고 생각했습니다.

저번에 반복문안에서 검증하는 방식이 올바르지 않은 방법인 것 같다고 말씀 해주셔서 검증 로직을 짜는게 생각보다 잘 안되었습니다.
2번째 요구사항 단위 테스트에서는 어떤식으로 검증하는 것이 좋을까요?
일단 저는 logger로 찍어 봤을때 싱글톤 객체로 생성되었다고 생각하고 있는 상태입니다.
어떤식으로 해결하면 좋을까요?

===============================================================================

준서님, 오늘 알려주신 방법 토대로 다시한번 만들어 봤습니다.

제가 여전히 어렵게 느끼는 부분이 Assertions를 사용한 테스트 코드 작성부분인데, 제가 머리가 굳어서 그런지 자꾸 for문안에서 해결하는 방법으로 코드가 가는 것 같습니다.

일단 2번째 요구사항 검증은 빈 팩토리 사이즈와 생성된 인스턴스의 사이즈가 동일하다면 성공하도록 짰습니다.

더 좋은 방법은 어떤 방법이 있을까요?

+ 추가로 커밋 메세지를 분리하는게 좋을 것 같다고 하셨는데 예를들어 테스트를 두개 작성할 때 커밋 메세지를 나눈다면 테스트 1번 작성하고 로컬에 커밋하고 2번 작성하고 로컬에 커밋해서 서로 분리를 시키면 좋을 것 같다는 말씀으로 이해 했는데 맞을까요?

kangminhyuk1111 committed:
리플렉션 실습 2단계

요구사항 2 문제점
- 요구사항 1에서 제시된 어노테이션들을 가져오는 것 까지 했으나 요구 사항 2의 문제를 해결할때 싱글톤 객체로 만들어야 해야하기 때문에 어떻게 해야 될지 방법을 찾아야함.
- 일단 구글링으로 싱글톤 패턴 구현하는 법 찾아봄

djawnstj reviewed (changes requested):
민혁님 전체적으로 이런 흐름으로 만들어주시면 돼요.

다만 지금 우리가 만드는 건 MVC 프레임워크에요. 프레임워크가 기본적으로 제공하는 DI 기능과 IoC 기능이 필요하고 이런 기능은 사용자가 이용할 수 있도록 해야 하니깐 `test` 패키지가 아닌 `main` 패키지에 작성을 해야 해요.
`main` 에 만든 프로덕션 코드가 정상적으로 작동 하는지 사용자 입장에서 확인 하는게 테스트 코드가 돼야 해요.

또한 이용자는 싱글톤 빈을 생성해야 하고, 이 역할을 프레임워크에게 위임 시키는 거에요.
그러면 프레임워크는 어떤 객체를 빈으로 생성하고 관리할 지 알아야 하는데, 이를 위해 필요한 것이 `Component`, `Controller`, `Service` 등등의 애너테이션이에요.
다시 말해 특정 애너테이션이 붙은 클래스는 프레임워크가 런타임에서 생성해 싱글톤으로 관리를 하면 됩니다.

스프링에서 애너테이션이 붙은 클래스를 별도로 생성하지 않아도 애플리케이션이 실행되면 해당 객체들이 존재했던 것과 같은 기능이에요.

그렇다면 만들어 주신 애너테이션들도 `main` 패키지에 존재하고 이용자들이 사용할 수 있도록 하면 될거 같습니다!

핵심은 `프레임워크가 부팅 시점에 동적으로 빈을 생성하고 생성된 빈 들은 런타임 동안 유지(관리)가 되어야 한다!` 입니다. 런타임을 강조한 이유를 고민해보시면 좋을거 같아요~


질문하신 내용에 대한 답을 드리자면,
1. 모든 애너테이션을 푸는게 맞을까요?
2. 싱글톤으로 생성했는지 검증을 어떻게 하나요?
3. 반복문을 돌려서 검증해야 하는 경우엔 어떻게 하나요?

인데 1번의 애너테이션을 푼다 라는 말을 잘 이해하지 못하겠어요😂
일단 관련된 내용으로 전하고 싶은 내용은 클래스 처럼 애너테이션도 일종의 상속 관계를 가질 수 있어요.
그렇게 되면 앞으로 `RestController`, `Properties` 등 여러 애너테이션이 추가 된다 하더라도 컴포넌트 스캔을 할 땐 하나의 컴포넌트 타입으로 모든 애너테이션에 대한 클래스 조회를 해올 수 있을거 같습니다!

2번은 자료구조 중 `Set` 이라는 자료구조를 이용하면 하나의 객체만 유일하게 생성할 수 있을거 같아요.
물론 Set 에게 타입 당 하나의 값만 가지도록 책임을 위임 하려면 동등성, 동일성 비교를 잘 해주어야 할 것 같습니다!
아니면 지금처럼 `Map` 자료구조에서 key-value 형식으로 구현해주셔도 될거 같아요.

일단 중요한 건 이미 생성된 적이 있는 클래스 타입인지 확인 후 생성된 객체가 존재한다면 예외를 던져야 할거 같아요.

3번은 지난 피드백에도 남겼지만, 검증하고자 하는 메서드는 분명 어떤 역할을 하고 무엇인가 상태값을 변경하거나 반환값을 가질 거에요.
지금 만든 `빈 등록 및 조회 테스트 - 모든 객체는 싱글톤 패턴으로 구현` 테스트는 `구현한 기능을 검증` 하는 역할이 아닌 어떤 기능 자체를 구현해버렸기 때문에 테스트가 길어지고 검증이 애매해지고 있어요.
테스트 코드는 `main()` 메서드로 사용하면 안된다고 생각하면 쉬울거에요.
일단 이 테스트에선 `instancesMap` 의 상태가 변할거처럼 보이는데요, 만약 이런 형태의 메서드를 검증하고자 한다면(예: `initInstanceMap(instanceMap)`) `instancesMap` 의 상태에 대한 검증을 하면 될 거 같습니다.

경우에 따라 만약 반환값이 여러개 생긴다면 배열이나 리스트를 통한 테스트 검증을 할 수 있고 자세한 방법은 노션 페이지에 공유해놨으니 확인 해보면 쉽게 할 수 있을거에요~

kangminhyuk1111 committed:
리플렉션 실습 2 - 로컬 push

kangminhyuk1111 committed:
리플렉션 실습 2단계

djawnstj commented on src/main/java/com/djawnstj/mvcframework/bean/BeanFactory.java:
```diff
@@ -0,0 +1,31 @@
+        reflections = new Reflections(basePackage);
+    }
+
+    public Set<Class<?>> scanAnnotation(Class<? extends Annotation> annotation) {
+        return reflections.getTypesAnnotatedWith(annotation)
+                .stream()
+                .filter(type -> !(type.isAnnotation() && type.isInterface()))
+                .collect(Collectors.toSet());
+    }
```
`BeanFactory` 객체가 클래스 정보를 scan 하는 역할까지 갖는게 맞을까요?
객체지향 설계원칙의 SRP 를 고려해보면 좋을거 같아요!

djawnstj commented on src/test/java/study/reflection/ComponentScannerTest.java:
```diff
@@ -1,82 +1,49 @@
-//        Assertions.assertTrue(annotatedClasses.contains(UserRepository.class));
-//        Assertions.assertTrue(annotatedClasses.contains(UserService.class));
+        // 두 컬렉션의 사이즈가 같다면 객체도 하나만 생성된 것이기 때문에 싱글톤
+        Assertions.assertEquals(classes.size(),instances.size());
```
`ComponentScanner` 의 이름을 보면 '컴포넌트를 스캔한다' 가 역할일거 같은데 인스턴스를 생성하는 테스트가 좋은 테스트인지 잘 모르겠어요😂
일단 `ComponentScanner` 라는 객체는 main 패키지의 프로덕션 코드로 존재해야 하지 않을까요?
그리고 해당 객체의 역할로 컴포넌트를 스캔하는 메서드가 만들어지고 해당 메서드를 테스트 하면 좋을거 같아요.

아마 `Array`, `List`, `Set`, `Map` 같은 자료구조 형태의 반환값이 나올거 같은데 이런 집합에 관한 테스트 검증은 노션에 `📖 테스트 코드` 를 참고하면 됩니다~

djawnstj reviewed (changes requested):
민혁님, `BeanFactory` 는 잘 만들어 주셨어요!
그런데 지금은 `BeanFactory` 의 역할이 조금 크기도 하고, 동적으로 생성한 빈이 정상적으로 생성되었는지 검증하는 테스트가 없어서 아쉬운거 같이요.

테스트에 대한 고민은 테스트 하고자 하는 메서드를 검증하는것이 첫번째에요.
그 메서드가 어떤 역할을 하고 어떤 결과를 만들어내는지 기대값이 있을텐데 
1. 반환값이 존재하면 우리가 기대한 반환값이 제대로 반환 되었는지
2. 파라미터로 전달한 참조 변수의 상태값을 변경하길 기대한다면 해당 상태값이 기대대로 변했는지
이런 부분을 검증하면 좋은 테스트 메서드일거 같아요.


저의 경우 커밋 메시지는 작업 단위로 묶으려고 해요.
컴포넌트 스캐너를 작업하면서 생긴 변경 단위, 빈팩토리를 작업하면서 생긴 변경 단위
이런식으로 묶으려고 합니다.
지금은 기존에 코드가 없어서 `BeanFactory` 를 작업하면 영향받는 외부 코드가 없을수도 있는데, 작엽량이 많아질수록 외부에서 코드가 변경되는 경우가 생길 수도 있어요.
그런 경우 하나의 커밋 단위로 묶습니다.

테스트 코드도 마찬가지로 변경한 작업에 영향을 받은 단위로 프로덕션 코드와는 별개의 커밋으로 묶어요.
만약 `BeanFactory` 를 새로 생성해서 테스트 메서드가 3개가 나왔다면 하나의 커밋 단위로 묶을거에요.
그런데 기존에 있던 `BeanFactory` 가 변경되어서 테스트 코드 클래스가 3개가 변경되었다면 이 3개 파일은 하나의 커밋으로 묶을거 같습니다.

피드백 보고 다시 작업해 주시고 만약 이해 안되거나 어려운 내용 있으면 슬랙으로 DM 주세요!

kangminhyuk1111 committed:
리플렉션 실습 2 - 로컬 push

kangminhyuk1111 committed:
리플렉션 실습 2 - 로컬 push

kangminhyuk1111 committed:
리플렉션 실습 2단계

kangminhyuk1111 committed:
리플렉션 실습 2단계

두번째 테스트가 싱글톤 패턴으로 구현이 되어있는지 검증할때 애매한 부분이 있는 것 같습니다. 애초에 ApplicationContext에서 중복검증을 하고 getBean을 해서 싱글톤으로 구현된 상태인 것 같은데 어떻게 억지로 같은 클래스지만 다른 객체를 넣어서 테스트 해볼수 있을까요? 제가 한 방법도 같은 하나의 객체를 가르키는지 테스트 하는 방법이라 틀린 것 같지는 않은데 다른 방법도 혹시나 있는지 궁금합니다.

kangminhyuk1111 committed:
리플렉션 실습 3단계

djawnstj commented on src/test/java/study/reflection/ComponentScannerTest.java:
```diff
@@ -14,29 +15,32 @@
+    @DisplayName("어노테이션 조회 테스트")
+    void testAnnotation() {
         Object bean1 = applicationContext.getBean(UserService.class);
         Object bean2 = applicationContext.getBean(UserRepository.class);
```
`ComponentScannerTest` 인데 `ApplicationContext` 가 테스트 대상 클래스네요.
역할에 알맞게 클래스 명, 패키지 명을 가져가는게 좋을거 같아요.

djawnstj commented on src/test/java/study/reflection/ComponentScannerTest.java:
```diff
@@ -14,29 +15,32 @@
 
+        Assertions.assertThat(bean1).isNotNull();
+        Assertions.assertThat(bean2).isNotNull();
+        Assertions.assertThat(bean1).isSameAs(bean2);
```
싱글톤 검증하는 방법으로 좋은거 같습니다!

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
`init` 메서드는 검증이 되지 않고있어요.

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -47,18 +47,20 @@ public void init() {
     }
 
-    public Object getBean(Class<?> clazz){
+    public Object getBean(Class<?> clazz) {
```
제네릭을 사용하지 않은 이유가 있을까요??

djawnstj commented on src/main/java/com/djawnstj/mvcframework/bean/BeanFactory.java:
`BeanFactory` 의 테스트 코드도 있으면 좋을거 같아요!

djawnstj reviewed (changes requested):
민혁님, 이제 테스트만 잘 다듬으면 2단계 마무리 하면 될거 같습니다!
싱글톤 검증은 지금 해주신 대로 검증해주시면 될거 같아요.

지금 빈을 생성하기 전 `if (!beans.containsKey(beanClass))` 로 검증을 하고 있는데, 이 조건은 절대 `false` 가 나올 수 없을거 같아요!
있으나 마나 한 조건인거 같은데 한번 고민해보시면 좋을거 같습니다.

kangminhyuk1111 committed:
리플렉션 실습 2단계

테스트 케이스 두개 추가 해봤습니다.
init을 테스트하면 BeforeEach문에 init이 빠져야되는게 맞나요?

djawnstj commented on src/test/java/study/reflection/ApplicationContextTest.java:
```diff
@@ -35,7 +44,9 @@ void testAnnotation() {
     @DisplayName("빈 등록 및 조회 테스트 - 모든 객체는 싱글톤 패턴으로 구현")
-    void addBeanTest() {
+    void addBeanSingletonTest() {
+        applicationContext.init();
```
`init` 메서드가 매번 호출 되어도 이미 `beanClasses` Set 에는 클래스들이 있기 때문에 변함이 없어요.
마찬가지로 `beans` Map 에도 변함이 없겠죠?

테스트를 작성할 때 중요한건 테스트 간에 영향이 있으면 안되요.
쉽게 말하면 테스트들은 서로 독립적인 상태여야 해요.

지금 상황에서 가장 쉬운 해결방법은 매 테스트 마다 `ApplicationContext` 를 생성해주는 방법이 있을거 같아요.

djawnstj commented on src/test/java/study/reflection/BeanFactoryTest.java:
```diff
@@ -0,0 +1,26 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(ReflectionTest.class);
+
+    @Test
+    @DisplayName("빈 팩토리 테스트")
+    void beanFactoryTest() {
+        BeanFactory beanFactory = new BeanFactory("com.djawnstj.mvcframework");
+
+        Set<Class<?>> classes = beanFactory.scanAnnotationClasses(Component.class);
+
+        Assertions.assertThat(classes).isNotNull();
+    }
```
`BeanFactoryTest` 잘 작성해 주었어요!
지금도 좋은데 굳이 아쉬운 점 하나 꼽자면 `Component.class` 들 만 스캔된게 맞을까?
이런 생각이 들거 같아요.

`BeanFactoryTest` 테스트 클래스 내부에 내부 클래스를 정의해 두고 
패키지 경로를 `BeanFactoryTest` 클래스로 잡으면 내가 원하는 클래스 개수가 조회된게 맞는지,
필요한 클래스만 조회된게 맞는지 검증할 수 있을거 같아요.

djawnstj reviewed (changes requested):
민혁님 거의 다 하셨어요!
다만 조금 아쉬운건 아직 테스트 코드 패키지가 제대로 잡혀있지 않고,
검증부 메서드가 여러개 있다는 점이에요.

제가 노션에 적어둔 대로 하나의 테스트에는 하나의 검증만 있어야 한다는 내용을 지켜보면 좋을거 같습니다.

몇가지 자잘한 수정만 하면 다음단계 진행하셔도 될거 같아요!

kangminhyuk1111 committed:
리플렉션 실습 2단계

실습 2단계 커밋 올렸습니다.
패키지 변경과 검증방식 변경했습니다.

djawnstj commented on src/test/java/study/bean/BeanFactoryTest.java:
```diff
@@ -1,4 +1,4 @@
-package study.reflection;
+package study.bean;
```
테스트 코드 패키지는 프로덕션 코드의 패키지와 동일하게 맞추는게 좋아요.

djawnstj commented on src/test/java/study/bean/BeanFactoryTest.java:
```diff
@@ -7,6 +7,7 @@
 import org.junit.jupiter.api.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import study.reflection.ReflectionTest;
```
테스트 코드와 상관 없는 클래스가 의존되고 있네요😂

복붙하거나 클래스명을 바꾸다가 이렇게 된거 같은데, 복붙할 땐 꼼꼼히 신경써야 해요.
클래스명이나 변수, 메서드 명 등을 변경할 땐 IDE 의 `Rename` 기능을 사용하는것을 추천드립니다!

djawnstj commented on src/test/java/study/context/ApplicationContextTest.java:
```diff
@@ -33,25 +30,29 @@ void testInit() {
     void annotationTest() {
+        ApplicationContext applicationContext = new ApplicationContext("com.djawnstj.mvcframework");
         applicationContext.init();
 
         Object bean1 = applicationContext.getBean(UserService.class);
         Object bean2 = applicationContext.getBean(UserRepository.class);
 
-        Assertions.assertThat(bean1).isNotNull();
-        Assertions.assertThat(bean2).isNotNull();
+        assertAll("annotationTest",
+                () -> Assertions.assertThat(bean1).isNotNull(),
+                () -> Assertions.assertThat(bean2).isNotNull());
```
피드백 반영 👍🏻

djawnstj reviewed (approved):
민혁님 고생하셨어요~
2단계 미션이 좀 길었는데, 또 가장 어렵기도 한 미션이라 앞으로는 비교적 빠르게 진행 할 수 있을거에요.

이번 단계는 여기서 머지하겠습니다!
